esphome/components/improv_serial/improv_serial_component.cpp

```diff
@@ -99,19 +99,19 @@ void ImprovSerialComponent::write_data_(std::vector<uint8_t> &data) {
 #endif  // !USE_ESP32_VARIANT_ESP32C3 && !USE_ESP32_VARIANT_ESP32S2 && !USE_ESP32_VARIANT_ESP32S3
       uart_write_bytes(this->uart_num_, data.data(), data.size());
       break;
-#if defined(CONFIG_ESP_CONSOLE_USB_CDC) && (defined(USE_ESP32_VARIANT_ESP32S2) || defined(USE_ESP32_VARIANT_ESP32S3))
+#ifdef USE_LOGGER_USB_CDC
```

Comment:
So I'm just realizing that, when this is set but `CONFIG_ESP_CONSOLE_USB_CDC` is not set, some code is excluded by IDF from the build, resulting in a linker error...

Reply:
Specifically, I believe this is the point of interest: https://github.com/espressif/esp-idf/blob/38eeba213aa695aabfd6d89aa9f5078dbe5a94c3/components/esp_system/port/soc/esp32s3/CMakeLists.txt#L19-L21
...so I think that's not being built, but were are still including code based on this define, which results in the linker error.

Reply:
I think the right move might be to do this:
```suggestion
#if defined(USE_LOGGER_USB_CDC) && defined(CONFIG_ESP_CONSOLE_USB_CDC)
```
@kpfleming any thoughts?

Reply:
We also need to add more Improv serial tests for each ESP that natively supports USB -- it probably would've caught the issue. 🤦🏻

Reply:
The linker error can be seen [here](https://github.com/esphome/esphome/actions/runs/9925278247/job/27417528461?pr=7081).
PR to fix is #7082 -- we can continue discussion there, if necessary.